File: frontend/src/resources/descriptors/node.ts
import { fetchGetNode, fetchGetNodes } from '@/services/general.service'
import { defineResource } from '../types'
import type { ResourceRef } from '../types'
import type { NodeRow, ResourceSummary } from '@/types/general.type'
import { model } from '../../../wailsjs/go/models'

export function nodeToRow(n: model.NodeDto): NodeRow {
	return {
		name: n.Name,
		memory: n.Resource.Memory,
		cpu: n.Resource.Cpu,
		storage: n.Resource.Storage,
		ephemeralStorage: n.Resource.StorageEphemeral,
		kubeletVersion: n.KubeletVersion,
		operatingSystem: n.OperatingSystem,
		version: n.Version,
		age: n.Age,
	}
}

export async function nodeToDetail(ref: ResourceRef): Promise<ResourceSummary> {
	const n = await fetchGetNode(ref.name, ref.cluster)
	return { name: n.Name, age: n.Age }
}

export const nodeDescriptor = defineResource({
	kind: 'node',
	namespaced: false,
	fetchList: fetchGetNodes,
	toRow: nodeToRow,
	fetchDetail: nodeToDetail,
})

File: internal/k8s/node.go
package k8s

import (
	"Kubexplorer/internal/model"
	"context"
	"fmt"
	"k8s.io/apimachinery/pkg/api/resource"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
)

type NodeClient struct {
	manager ClusterResolver
}

func NewNode(manager ClusterResolver) *NodeClient {
	return &NodeClient{manager: manager}
}

func (n *NodeClient) GetNodes(ctx context.Context, clusterCtx string) ([]model.NodeDto, error) {
	client, err := n.manager.ResolveClusterContext(clusterCtx)
	if err != nil {
		return nil, fmt.Errorf("kubeclient: error resolving cluster context: %v", err)
	}

	nodes, err := client.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
	if err != nil {
		return nil, fmt.Errorf("listing nodes: %w", err)
	}

	var result []model.NodeDto

	for _, node := range nodes.Items {
		dto := model.NodeDto{
			Name: node.Name,
			Resource: model.Resource{
				Cpu:     node.Status.Capacity.Cpu().Value(),
				Memory:  node.Status.Capacity.Memory().ScaledValue(resource.Giga),
				Storage: node.Status.Capacity.StorageEphemeral().ScaledValue(resource.Giga),
			},
			KubeletVersion:  node.Status.NodeInfo.KubeletVersion,
			OperatingSystem: node.Status.NodeInfo.OperatingSystem,
			Version:         node.ResourceVersion,
			Age:             model.FormatAge(node.CreationTimestamp.Time),
			CreatedAt:       node.CreationTimestamp.Unix(),
			Labels:          node.Labels,
		}

		result = append(result, dto)
	}

	return result, nil
}

func (n *NodeClient) GetNode(ctx context.Context, ref model.ResourceRef) (model.NodeDto, error) {
	name, clusterCtx := ref.Name, ref.Cluster
	client, err := n.manager.ResolveClusterContext(clusterCtx)
	if err != nil {
		return model.NodeDto{}, fmt.Errorf("kubeclient: error resolving cluster context: %v", err)
	}

	node, err := client.CoreV1().Nodes().Get(ctx, name, metav1.GetOptions{})
	if err != nil {
		return model.NodeDto{}, fmt.Errorf("getting node %s: %w", name, err)
	}

	return model.NodeDto{
		Name: node.Name,
		Resource: model.Resource{
			Cpu:     node.Status.Capacity.Cpu().Value(),
			Memory:  node.Status.Capacity.Memory().ScaledValue(resource.Giga),
			Storage: node.Status.Capacity.StorageEphemeral().ScaledValue(resource.Giga),
		},
		Version:   node.ResourceVersion,
		Age:       model.FormatAge(node.CreationTimestamp.Time),
		CreatedAt: node.CreationTimestamp.Unix(),
		Labels:    node.Labels,
	}, nil
}
